### PR TITLE
ParameterJuMP integration

### LINKED ISSUE #22: Write code for time series update in simulations

### LINKED ISSUE #99: Handle initial conditions and time series with ParameterJuMP.jl

### PR DESCRIPTION
**Pull request by jd-lara**
Parameters enter in an Operational Model in three places. 
1. RH of `<=` and `>=` constraints. 
2. Nodal injections in the `Expressions` field of canonical_model
3. Initial conditions of devices. This required a significant change to handle initial conditions properly. 

This PR addresses the integration of `ParameterJuMP` into PowerSimulations to facilitate sequential solving to address #109 and fixes #22 and closes #99 

As soon as the new `ParameterJuMP` branches are merged, I will update the manifest. 

Please see the comment in #109 about the Duration Constraint.

### CODE REVIEW

**Comment by jd-lara** [comment]
@jmaack24 I just pushed 9538fcb that enables `ParameterJuMP` with `AbstractVariableRef` Let me know if this works with the `StructJuMP` implementation you have been working on.

**Comment by jd-lara** [comment]
@jmaack24 this can be addressed here. 

https://github.com/NREL/PowerSimulations.jl/blob/9031bd43f026f695fffa73bbf94f5d2f6f340967/src/core/operations_constructor.jl#L1-L18

Running `PJ.enable_parameters` on a StructJuMPModel

**Comment by jmaack24** [comment]
> @jmaack24 this can be addressed here.
> 
> [PowerSimulations.jl/src/core/operations_constructor.jl](https://github.com/NREL/PowerSimulations.jl/blob/9031bd43f026f695fffa73bbf94f5d2f6f340967/src/core/operations_constructor.jl#L1-L18)
> 
> Lines 1 to 18 in [9031bd4](/NREL/PowerSimulations.jl/commit/9031bd43f026f695fffa73bbf94f5d2f6f340967)
> 
>  function _pass_abstract_jump(optimizer::Union{Nothing,JuMP.OptimizerFactory}; kwargs...) 
>   
>      if isa(optimizer,Nothing) 
>          @info("The optimization model has no optimizer attached") 
>      end 
>   
>      if :JuMPmodel in keys(kwargs) 
>   
>          return kwargs[:JuMPmodel] 
>   
>      end 
>   
>      JuMPmodel = JuMP.Model(optimizer) 
>      PJ.enable_parameters(JuMPmodel) 
>   
>      return JuMPmodel 
>   
>  end 
> Running `PJ.enable_parameters` on a StructJuMPModel

Except ParameterJuMP that function is only implemented for JuMP.Model type and not any subtype of JuMP.AbstractModel:
`function enable_parameters(m::JuMP.Model)`
The PJ code does this in numerous places.

**Comment by jd-lara** [comment]
Ok, I will make the feature request in ParameterJuMP. It shouldn't take a long time actually.

**Comment by jd-lara** [comment]
> Ok, I will make the feature request in ParameterJuMP. It shouldn't take a long time actually. In the meantime, if you have an example that I can use for testing, I could be working on a PR to ParameterJuMP to address these changes.

I think this is going to be resolved in StructJuMP and ParameterJuMP so we can close this conversation

File: src/devices/device_models/thermal_generation.jl [hunk 1]
```diff
@@ -184,44 +203,95 @@ end
...
+            ramp_params[i] = g.tech.ramplimits
+            minmax_params[i] = g.tech.activepowerlimits
+            i, state = iterate(idx, state)
+            state === nothing && (i += 1; break)                                                                   
+        end 
+    end
+    
+    deleteat!(set_name, i:last(idx))
+    deleteat!(ramp_params, i:last(idx))
+    deleteat!(minmax_params, i:last(idx))
```
Comment (claytonpbarrows) [inline src/devices/device_models/thermal_generation.jl]: These make the T+1 time period nonbinding?

Comment (jd-lara) [reply src/devices/device_models/thermal_generation.jl]: Line 220 adds a +1 to i to avoid that

**Pull request merged**